include/lbann/layers/activations/softmax.hpp

```diff
@@ -34,411 +32,335 @@
-    m_min_output = other.m_min_output;
+  softmax_layer(const softmax_layer& other)
+    : activation_layer(other),
+      m_min_output(other.m_min_output) {
+
+    // Matrix deep copy
+    m_workspace = other.m_workspace;
+    if (m_workspace != nullptr) { m_workspace = m_workspace->Copy(); }
+
+    // Copy GPU objects
     this->m_cudnn = other.m_cudnn;
     this->m_using_gpus = other.m_using_gpus;
-#if defined(LBANN_HAS_CUDA) && defined(LBANN_SOFTMAX_CUDA)
-    if (this->m_using_gpus) {
-      cudnn::copy_tensor_cudnn_desc(other.m_cudnn_desc,
-                                    m_cudnn_desc);
-    }
-#endif
+
   }
 
   softmax_layer& operator=(const softmax_layer& other) {
     activation_layer::operator=(other);
-    if (m_workspace) {
-      delete m_workspace;
-      delete m_workspace_v;
-    }
-    m_workspace = other.m_workspace->Copy();
-    m_workspace_v = other.m_workspace_v->Copy();
     m_min_output = other.m_min_output;
+
+    // Deep matrix copy
+    if (m_workspace != nullptr) { delete m_workspace; }
+    m_workspace = other.m_workspace;
+    if (m_workspace != nullptr) { m_workspace = m_workspace->Copy(); }
+
+    // Copy GPU objects
     this->m_cudnn = other.m_cudnn;
     this->m_using_gpus = other.m_using_gpus;
-#if defined(LBANN_HAS_CUDA) && defined(LBANN_SOFTMAX_CUDA)
-    if (this->m_using_gpus) {
-      cudnn::copy_tensor_cudnn_desc(other.m_cudnn_desc,
-                                    m_cudnn_desc);
-    }
-#endif
+
   }
 
   ~softmax_layer() override {
-    delete m_workspace;
-    delete m_workspace_v;
-
-#if defined(LBANN_HAS_CUDA) && defined(LBANN_SOFTMAX_CUDA)
-    if (m_cudnn_desc) {
-      CHECK_CUDNN(cudnnDestroyTensorDescriptor(m_cudnn_desc));
-    }
-#endif
-
+    if (m_workspace != nullptr) { delete m_workspace; }
   }
 
   softmax_layer* copy() const override { return new softmax_layer(*this); }
-
   std::string get_type() const override { return "softmax"; }
 
   std::string get_description() const override {
     return std::string {} + " softmax" + " dataLayout: "
            + this->get_data_layout_string(get_data_layout());
   }
 
-  inline void initialize_distributed_matrices() override;
   data_layout get_data_layout() const override { return T_layout; }
 
-  void setup_data() override {
-    activation_layer::setup_data();
-    m_workspace->Resize(1, this->m_model->get_max_mini_batch_size());
+  void setup_matrices(const El::Grid& grid) override {
+    activation_layer::setup_matrices(grid);
+    if (m_workspace != nullptr) { delete m_workspace; }
+    switch (get_data_layout()) {
+    case data_layout::MODEL_PARALLEL:
+      m_workspace = new StarMRMat(grid);
+      break;
+    case data_layout::DATA_PARALLEL:
+      m_workspace = new StarVCMat(grid);
+      break;
+    default:
+      std::stringstream err;
+      err << __FILE__ << " " << __LINE__ << " :: "
+          << "invalid distributed matrix layout";
+      throw lbann_exception(err.str());
+    }
```

Comment:
Is this data allocation going to be updated to the new version that we discussed, using templated functions for each data layout.